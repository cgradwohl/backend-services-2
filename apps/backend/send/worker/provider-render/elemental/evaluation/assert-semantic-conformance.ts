import { ElementalIR, ElementalNodeIR } from "~/api/send/types";
import { ElementalError } from "../errors";
import { ElementalRefs } from "./types";

export function assertSemanticConformance(
  ir: ElementalIR,
  refTable: ElementalRefs,
  /** Internal use only */
  startIndex = 0
): void {
  ir.forEach((element, index) => {
    const totalIndex = startIndex + index;

    // Analyze element if statement
    if ("if" in element) {
      const refIds = extractRefIdsFromString(element["if"]);

      refIds.forEach((ref) => {
        const referencedElement = checkForUseOfRefThatDoesNotExist(
          ref,
          refTable
        );
        checkForUseBeforeDefined(totalIndex, referencedElement);
      });
    }

    // Handle children
    if ("elements" in element) {
      assertSemanticConformance(element.elements, refTable, totalIndex);
    }
  });
}

/** Throws an error if ref is invalid, otherwise returns the referenced element */
export function checkForUseOfRefThatDoesNotExist(
  refId: string,
  refs: ElementalRefs
) {
  const referencedElement = refs[refId];
  if (!referencedElement) {
    throw new ElementalError(`Element with ref "${refId}" does not exist.`);
  }

  return referencedElement;
}

// Also covers circular references
export function checkForUseBeforeDefined(
  currentElementIndex: number,
  referencedElement: ElementalNodeIR
) {
  if (referencedElement.index > currentElementIndex) {
    throw new ElementalError(
      `Element references "${referencedElement.ref}" before it is defined.`
    );
  }
}

/** Returns an array of all refIds found in the string */
export function extractRefIdsFromString(str: string): string[] {
  // Scan string for refs in the format of refs.refName
  const refs = str.match(/refs\.([a-zA-Z0-9_]+)/g) ?? [];
  return refs.map((ref) => ref.slice(5));
}
