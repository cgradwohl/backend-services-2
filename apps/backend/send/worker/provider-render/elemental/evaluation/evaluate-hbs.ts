import handlebars from "handlebars";
import { ElementalNode, ElementalNodeIR } from "~/api/send/types";
import { HandlebarsEvalError } from "~/lib/jsonnet/jsonbars";

export const supportedFields = ["content", "title", "href", "src"] as const;

export function evaluateHbsOfElement(
  element: ElementalNodeIR,
  data: any
): ElementalNodeIR {
  const copy = { ...element };

  supportedFields.forEach((field) => {
    if (field in copy) {
      copy[field] = render(copy[field], data);
    }
  });

  return copy;
}

export function elementHasHbsSupportedField(element: ElementalNode): boolean {
  return Object.keys(element).some((key) =>
    (supportedFields as ReadonlyArray<string>).includes(key)
  );
}

function render(content: string, data: any): string {
  try {
    const template = handlebars.compile(content, { noEscape: true });
    return template(data);
  } catch (error: any) {
    throw new HandlebarsEvalError(error.message);
  }
}
