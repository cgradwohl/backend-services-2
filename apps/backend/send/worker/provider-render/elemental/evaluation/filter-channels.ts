import { ElementalIR, ElementalChannelNodeIR } from "~/api/send/types";
import { ElementalError } from "../errors";

/**
 * Evaluates channel requirements of an element and filters the ir based on the results.
 * Must be done before other evaluation. Does not update visibility of refs.
 */
export function filterChannels({
  ir,
  channel,
}: {
  ir: ElementalIR;
  channel?: string;
}): ElementalIR {
  // Handle top level channel elements
  if (ir.some((element) => element.type === "channel")) {
    return filterTopLevelChannelElements({ ir, channel });
  }

  return filterChannelSpecificElements({ ir, channel });
}

export function filterChannelSpecificElements({
  ir,
  channel,
}: {
  ir: ElementalIR;
  channel?: string;
}): ElementalIR {
  return ir
    .filter((element) => {
      if ("channels" in element) {
        return element.channels.includes(channel);
      }

      return true;
    })
    .map((element) => {
      if ("elements" in element) {
        element.elements = filterChannelSpecificElements({
          ir: element.elements,
          channel,
        });
      }

      return element;
    });
}

export function filterTopLevelChannelElements({
  ir,
  channel = "default",
}: {
  ir: ElementalIR;
  channel?: string;
}): ElementalIR {
  if (!ir.every((element) => element.type === "channel")) {
    throw new ElementalError(
      "All top level elements must be channels unless no channel element is present."
    );
  }

  const exactChannelMatch = ir.find(
    (element: ElementalChannelNodeIR) => element.channel === channel
  ) as ElementalChannelNodeIR;

  if (exactChannelMatch) {
    return [exactChannelMatch];
  }

  const defaultChannelMatch = ir.find(
    (element: ElementalChannelNodeIR) => element.channel === "default"
  ) as ElementalChannelNodeIR;

  if (defaultChannelMatch) {
    return [defaultChannelMatch];
  }

  throw new ElementalError(
    `No channel element found for taxonomy: ${channel}. Consider specifying a default.`
  );
}
