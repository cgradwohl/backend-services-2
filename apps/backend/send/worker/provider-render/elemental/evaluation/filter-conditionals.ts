import { VM } from "vm2";
import { ElementalIR, ElementalNodeIR, MessageData } from "~/api/send/types";
import { IProfile } from "~/types.api";
import { ElementalError } from "../errors";
import { ElementalRefs } from "./types";

export interface FilterConditionalsOpts {
  ir: ElementalIR;
  data: MessageData;
  refs: ElementalRefs;
  profile: IProfile;
}

/**
 * Filters out elements that fail their conditional check.
 *
 * Note. Unlike filterChannels, filterConditionals updates the visible property
 * of the element on the ref table.
 */
export function filterConditionals({
  ir,
  data,
  refs,
  profile,
}: FilterConditionalsOpts): ElementalIR {
  const vm = new VM({ sandbox: { data, refs, profile } });
  return runConditionalEvaluation(ir, vm);
}

function runConditionalEvaluation(ir: ElementalIR, vm: VM): ElementalIR {
  return ir
    .filter((element) => {
      if (!element["if"]) {
        return true;
      }

      const ifResult = vm.run(element["if"]);
      if (typeof ifResult !== "boolean") {
        throw new ElementalError(
          "A conditional expression must evaluate to a boolean."
        );
      }

      markElementVisibility(element, ifResult, vm.sandbox.refs);
      return ifResult;
    })
    .map((element) => {
      if ("elements" in element) {
        return {
          ...element,
          elements: runConditionalEvaluation(element.elements, vm),
        };
      }

      return element;
    });
}

function markElementVisibility(
  element: ElementalNodeIR,
  visible: boolean,
  refs: ElementalRefs
) {
  if (visible) {
    return;
  }

  const refElement = refs[element.ref];
  if (refElement) {
    refElement.visible = false;
  }
}
