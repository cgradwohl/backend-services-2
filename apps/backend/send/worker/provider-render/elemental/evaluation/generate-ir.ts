import { ElementalNodeIR, ElementalNode, ElementalIR } from "~/api/send/types";
import { ElementalRefs } from "./types";

/** Fills out the metadata of an element and returns the refTable. Please suggest a better name. */
export function generateIR(
  elements: ElementalNode[],
  refs: ElementalRefs = {},
  startIndex = 0
): ElementalIR {
  return elements
    .filter((element) => element.type !== "comment")
    .map((element, index) => {
      const totalIndex = startIndex + index;
      return {
        ...element,
        index: totalIndex,
        visible: true,
        ...("elements" in element
          ? { elements: generateIR(element.elements, refs, totalIndex + 1) }
          : undefined),
      } as ElementalNodeIR;
    });
}
