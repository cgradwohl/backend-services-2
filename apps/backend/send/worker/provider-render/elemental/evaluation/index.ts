export { evaluateElemental } from "./evaluate-elemental";
