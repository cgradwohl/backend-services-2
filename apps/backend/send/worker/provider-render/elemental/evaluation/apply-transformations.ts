import {
  ElementalIR,
  ElementalNodeIR,
  MessageData,
  UTM,
} from "~/api/send/types";
import { pipe } from "~/lib/pipe";
import { evaluateHbsOfElement } from "./evaluate-hbs";
import { addUtmToElementalHref } from "../../augment-href";
import DOMPurify from "isomorphic-dompurify";
import { IProfile } from "~/types.api";

export function applyTransformations(opts: {
  ir: ElementalIR;
  data: any;
  profile: IProfile;
  utm?: UTM;
}): ElementalIR {
  const { ir, ...context } = opts;
  return ir.map((element) => {
    const evaluated = applyFieldTransformations({ element, ...context });

    if ("elements" in evaluated) {
      evaluated.elements = applyTransformations({
        ir: evaluated.elements,
        ...context,
      });
    }

    return evaluated;
  });
}

export function applyFieldTransformations({
  element,
  data,
  profile,
  utm,
}: {
  element: ElementalNodeIR;
  data: MessageData;
  profile: IProfile;
  utm?: UTM;
}): ElementalNodeIR {
  return pipe(element)
    .into((element) => evaluateHbsOfElement(element, { profile, ...data }))
    .into((element) => addUtmToElementalHref(element, utm))
    .into(transformHtml)
    .complete();
}

export function transformHtml(element: ElementalNodeIR): ElementalNodeIR {
  if (element.type !== "html") {
    return element;
  }

  return {
    ...element,
    content: DOMPurify.sanitize(element.content, { FORCE_BODY: true }),
  };
}
