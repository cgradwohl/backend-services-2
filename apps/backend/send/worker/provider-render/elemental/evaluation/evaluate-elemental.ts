import {
  MessageMetadata,
  ElementalIR,
  ElementalNode,
  MessageData,
  UTM,
} from "~/api/send/types";
import { filterChannels } from "./filter-channels";
import { filterConditionals } from "./filter-conditionals";
import { generateIR } from "./generate-ir";
import { generateRefs } from "./generate-refs";
import { evaluateLoops } from "./loop-evaluation";
import { assertSemanticConformance } from "./assert-semantic-conformance";
import { updateVisibility } from "./update-visibility";
import { pipe } from "~/lib/pipe";
import { interpolateLocales } from "./interpolate-locales";
import { applyTransformations } from "./apply-transformations";
import { IProfile } from "~/types.api";

type ElementEvaluator = {
  channel?: string;
  elements: ElementalNode[];
  locale?: string;
  data: MessageData;
  utm?: UTM;
  profile: IProfile;
};

/**
 * Evaluates the control flow of the elements and produces an IR with channels and failed
 * conditionals filtered out. It also renders the looped elements
 */
export function evaluateElemental({
  elements,
  channel,
  data,
  locale,
  utm,
  profile,
}: ElementEvaluator): ElementalIR {
  const ir = generateIR(elements);
  const refs = generateRefs(ir);
  assertSemanticConformance(ir, refs);

  return pipe(ir)
    .into((ir) => filterChannels({ ir, channel }))
    .into((ir) => updateVisibility(ir, refs))
    .into((ir) => filterConditionals({ ir, data, profile, refs }))
    .into((ir) => interpolateLocales({ ir, locale }))
    .into((ir) => evaluateLoops({ ir, data, profile, refs }))
    .into((ir) => applyTransformations({ ir, data, profile, utm }))
    .complete();
}
