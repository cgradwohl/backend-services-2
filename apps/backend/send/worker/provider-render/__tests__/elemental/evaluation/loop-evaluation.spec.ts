import { generateIR } from "../../../elemental/evaluation/generate-ir";
import {
  evaluateLoops,
  evaluateLoop,
  assertIsValidLoopResult,
} from "../../../elemental/evaluation/loop-evaluation";

describe("loop evaluation", () => {
  describe("evaluateLoops", () => {
    it("should evaluate a basic loop", () => {
      const ir = generateIR([
        {
          type: "group",
          elements: [
            {
              type: "text",
              content: "{{ add $.index 1 }}. Hello {{ $.item }}!",
              loop: "data.names",
            },
          ],
        },
      ]);
      const data = {
        names: ["Drew", "Chris", "Suhas"],
      };
      const evaluated = evaluateLoops({ ir, data, refs: {}, profile: {} });
      expect(evaluated).toMatchSnapshot();
    });

    it("should evaluate a complex loop with groups and recursion", () => {
      const ir = generateIR([
        {
          type: "group",
          loop: "data.products.desserts",
          elements: [
            {
              type: "text",
              content: "# Name: {{$.item.name}}, Price: {{$.item.price}}",
            },
            {
              type: "text",
              content: "Flavors:",
            },
            {
              type: "text",
              content: "- {{$.item}}",
              loop: "$.item.flavors",
            },
          ],
        },
      ]);

      const data = {
        products: {
          desserts: [
            {
              name: "Ice Cream",
              price: "$9.00",
              flavors: ["Chocolate", "Vanilla"],
            },
            {
              name: "Cake",
              price: "$9.00",
              flavors: ["German Chocolate", "Cheesecake"],
            },
          ],
        },
      };

      const evaluated = evaluateLoops({ ir, data, refs: {}, profile: {} });
      expect(evaluated).toMatchSnapshot();
    });

    it("should interpolate variables outside of content", () => {
      const ir = generateIR([
        {
          type: "action",
          content: "View {{ $.item.name }}",
          href: "{{$.item.href}}",
          loop: "data.products",
        },
      ]);
      const data = {
        products: [
          {
            name: "Cheese",
            href: "https://www.cheese.com",
          },
          {
            name: "Bread",
            href: "https://www.bread.com",
          },
        ],
      };
      const evaluated: any = evaluateLoops({ ir, data, refs: {}, profile: {} });
      expect(evaluated[0].content).toBe("View Cheese");
      expect(evaluated[0].href).toBe("https://www.cheese.com");
      expect(evaluated[1].content).toBe("View Bread");
      expect(evaluated[1].href).toBe("https://www.bread.com");
    });
  });

  describe("evaluateLoop", () => {
    it("should evaluate loops", () => {
      const data = {
        products: [
          { name: "Ice Cream!", price: 1.99 },
          { name: "Cake!", price: 2.99 },
          { name: "Cleveland (The City) -_-", price: 0.99 },
        ],
      };
      const loop = "data.products";

      const evaluated = evaluateLoop({ refs: {}, data, loop, profile: {} });
      expect(evaluated.length).toBe(3);
      expect(evaluated).toEqual(data.products);
    });

    it("should throw when loop does not return array", () => {
      const data = {
        boop: 1,
      };
      const loop = "data.boop";
      expect(() =>
        evaluateLoop({ refs: {}, data, loop, profile: {} })
      ).toThrow();
    });
  });

  describe("assertIsValidLoopResult", () => {
    it("should throw in typeof loopResult is not an array", () => {
      const loopResult = {};
      expect(() => {
        assertIsValidLoopResult(loopResult);
      }).toThrow();
    });

    it("should not throw in typeof loopResult is an array", () => {
      const loopResult = [];
      expect(() => {
        assertIsValidLoopResult(loopResult);
      }).not.toThrow();
    });
  });
});
