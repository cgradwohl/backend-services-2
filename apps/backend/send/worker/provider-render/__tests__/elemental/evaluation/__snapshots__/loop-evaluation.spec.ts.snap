// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loop evaluation evaluateLoops should evaluate a basic loop 1`] = `
Array [
  Object {
    "elements": Array [
      Object {
        "content": "1. Hello Drew!",
        "index": 1,
        "type": "text",
        "visible": true,
      },
      Object {
        "content": "2. Hello Chris!",
        "index": 1,
        "type": "text",
        "visible": true,
      },
      Object {
        "content": "3. Hello Suhas!",
        "index": 1,
        "type": "text",
        "visible": true,
      },
    ],
    "index": 0,
    "type": "group",
    "visible": true,
  },
]
`;

exports[`loop evaluation evaluateLoops should evaluate a complex loop with groups and recursion 1`] = `
Array [
  Object {
    "elements": Array [
      Object {
        "content": "# Name: Ice Cream, Price: $9.00",
        "index": 1,
        "type": "text",
        "visible": true,
      },
      Object {
        "content": "Flavors:",
        "index": 2,
        "type": "text",
        "visible": true,
      },
      Object {
        "content": "- Chocolate",
        "index": 3,
        "type": "text",
        "visible": true,
      },
      Object {
        "content": "- Vanilla",
        "index": 3,
        "type": "text",
        "visible": true,
      },
    ],
    "index": 0,
    "type": "group",
    "visible": true,
  },
  Object {
    "elements": Array [
      Object {
        "content": "# Name: Cake, Price: $9.00",
        "index": 1,
        "type": "text",
        "visible": true,
      },
      Object {
        "content": "Flavors:",
        "index": 2,
        "type": "text",
        "visible": true,
      },
      Object {
        "content": "- German Chocolate",
        "index": 3,
        "type": "text",
        "visible": true,
      },
      Object {
        "content": "- Cheesecake",
        "index": 3,
        "type": "text",
        "visible": true,
      },
    ],
    "index": 0,
    "type": "group",
    "visible": true,
  },
]
`;
